src/util/metadata/header_metadata.rs

```diff
@@ -402,12 +402,12 @@ impl HeaderMetadataSpec {
         #[cfg(debug_assertions)]
         self.assert_spec::<T>();
         if self.num_of_bits < 8 {
-            let (_, mask) = self.get_shift_and_mask_for_bits();
-            let new_val = val.to_u8().unwrap() | !mask;
+            let (lshift, mask) = self.get_shift_and_mask_for_bits();
+            let new_val = (val.to_u8().unwrap() | !mask) << lshift;
```

Comment:
`mask` is shifted to the high order bits in 
https://github.com/mmtk/mmtk-core/blob/53b6d57c2a96f6064e01e245ea0be05dd673cfad/src/util/metadata/header_metadata.rs#L85, so `mask` should be used with the high order bits. `val` is the value on the low order bits. So it is incorrect to bitor `val` with `!mask`.

Probably we should have `(val.to_u8().unwrap() << lshift) | !mask`.

Reply:
I think it would be good if you can add at least one test for this.

Reply:
Yes.  I changed as you suggested.